pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,147 @@
+import logging
+import datetime
+import json
+import re
+
+import jsonschema
+
+from pubtools.pulplib._impl import compat_attr as attr
+from ..schema import load_schema
+from .common import InvalidDataException
+
+
+LOG = logging.getLogger("pubtools.pulplib")
+
+
+def _iso_time_now():
+    return datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
+
+
+@attr.s(kw_only=True)
+class MaintainedRepo(object):
+    """Details about the maintenance"""
+
+    message = attr.ib(default=None, type=str)
+    """Why this repository is in maintenance"""
+    owner = attr.ib(default=None, type=str)
+    """Who set this repository in maintenance mode"""
+    started = attr.ib(default=None, type=str)
+    """When did the maintenance start"""
+
+
+@attr.s(kw_only=True)
+class MaintenanceReport(object):
+    """Represent the maintenance status of pulp"""
+
+    _SCHEMA = load_schema("maintenance")
+
+    last_updated = attr.ib(default=_iso_time_now(), type=str)
+    """:class:`~datetime.datetime` in UTC when this report was last updated,
+    if it's the first time the report is created, current time is used."""
+
+    last_updated_by = attr.ib(default="Content Delivery", type=str)
+    """Person/party who updated the report last time"""
+
+    repos = attr.ib(default={}, type=dict)
```

Comment:
I know that these are stored as a dict on the underlying data format, but I'd suggest not doing that here.  Mainly because it gets you this weird asymmetric data structure, as mentioned in an earlier comment, where each element in the dict carries all info about the maintenance mode entry *except* the repo ID itself, so you have to pass that around separately.

I suggest a MaintenanceEntry class for each item in this container, and that class should have a repo ID, and this should be a tuple of entries rather than a dict.  Why tuple: because those are immutable.

```suggestion
    entries = attr.ib(default=attr.Factory(tuple), type=tuple)
    """A list of maintenance mode entries.  Each entry contains metadata for a single repository
    currently in maintenance mode."""
```

To keep the invariant that there are no duplicate entries for a single repo, a validator could be implemented.

Also incidentally, not 100% sure about this, but I guess setting default={} with attrs is unsafe for the same reason as setting some default={} as a function/method argument is unsafe - presumably all the instances get a reference to the *same* dict, which could be modified and would then affect many objects.

Reply:
for the second point about default to {} is unsafe because it might refer to the same object, I tested a little and seems like it's safe to use:
`>>> import attr
>>> @attr.s
... class c(object):
...     c = attr.ib(default={}, type=dict)
... 
>>> c
<class '__main__.c'>
>>> c()
c(c={})
>>> c(c={'a':'b'})
c(c={'a': 'b'})
>>> c()
c(c={})
`

Reply:
The unsafe thing would be when you modify the contents of that dict.  I did my own testing just now and indeed it seems to be unsafe in that every instance shares the same dict, probably not intended:

```
>>> import attr
>>> @attr.s
... class Klass:
...   c = attr.ib(default={})
... 
>>> k1 = Klass()
>>> k2 = Klass()
>>> k1.c
{}
>>> k2.c
{}

# Now let's try assigning something to the dict

>>> k1.c['foo'] = 'bar'
>>> k1.c
{'foo': 'bar'}

# Oops, my change on k1 is also visible via k2
>>> k2.c
{'foo': 'bar'}
```